nexus/src/app/background/metrics_producer_gc.rs

```diff
@@ -0,0 +1,116 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for garbage collecting metrics producers that have not
+//! renewed their lease
+
+use super::common::BackgroundTask;
+use chrono::TimeDelta;
+use chrono::Utc;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use serde_json::json;
+use slog_error_chain::InlineErrorChain;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Background task that prunes metrics producers that have failed to renew
+/// their lease.
+pub struct MetricProducerGc {
+    datastore: Arc<DataStore>,
+    lease_duration: Duration,
+}
+
+impl MetricProducerGc {
+    pub fn new(datastore: Arc<DataStore>, lease_duration: Duration) -> Self {
+        Self { datastore, lease_duration }
+    }
+
+    async fn activate(&mut self, opctx: &OpContext) -> serde_json::Value {
+        // TODO We should turn this task on as a part of landing the rest of the
+        // move to metric producer leases. For now, we leave it disabled to
+        // avoid pruning producers that don't know to renew leases.
+        if true {
+            warn!(
+                opctx.log,
+                "Metric producer GC: statically disabled pending omicron#5284"
+            );
+            return json!({
+                "error": "metric producer gc disabled (omicron#5284)",
+            });
+        }
+
+        let Some(expiration) = TimeDelta::from_std(self.lease_duration)
+            .ok()
+            .and_then(|delta| Utc::now().checked_add_signed(delta))
```

Comment:
Should this be a checked subtraction? I might be missing something, but this is always in the future since the lease duration is positive, right?

Reply:
... yes, yes it should. And it looks like I need at least a basic test here of `activate` to catch this.

Reply:
Test added (and confirmed this should be `sub`) in [7e60b54](https://github.com/oxidecomputer/omicron/pull/5311/commits/7e60b54df48846c1265fdde17944d8214adc132e)